### PR TITLE
convert tests from JUnit3 to JUnit4

### PR DESCRIPTION
**Pull request by darxriggs**
- JUnit3 is quite old and should be replaced by JUnit4 (or maybe even spock)
- for now touching all tests that directly inherit from junit.framework.TestCase
- tests with other base classes may follow

### CODE REVIEW

**Comment by darxriggs** [comment]
@jglick I saw you also performed some JUnit3 to JUnit4 conversion via babfb832a39725b6b3c8edc4d49a58c7549f518d.

Could you maybe have look at this pull request?

File: core/src/test/java/hudson/BulkChangeTest.java [hunk 1]
```diff
@@ -1,18 +1,18 @@
 /*
  * The MIT License
- * 
+ *
```
Comment (jglick) [inline core/src/test/java/hudson/BulkChangeTest.java]: Please avoid gratuitous (for example, whitespace-only) changes in diffs in PRs. They make the diff much harder to read, increase the change of merge conflicts, and cause lots of trouble down the road when cherry-picking, running `git-blame`, etc.

File: core/src/test/java/hudson/MarkupTextTest.java [hunk 2]
```diff
@@ -102,7 +112,8 @@ public void testPreEscape() {
     }
 
     /* @Bug(6252) */
```
Comment (jglick) [inline core/src/test/java/hudson/MarkupTextTest.java]: Feel free to replace any such occurrences with e.g. `@Issue("JENKINS-6252")`.

File: core/src/test/java/hudson/util/ConsistentHashTest.java [hunk 3]
```diff
@@ -135,7 +141,6 @@ public void testSpeed() {
         for (int j=0; j<10; j++) {
             ConsistentHash<String> b = new ConsistentHash<String>();
             b.addAll(data);
-//            System.out.println(Iterables.toString(b.list("x")));
```
Comment (jglick) [inline core/src/test/java/hudson/util/ConsistentHashTest.java]: Can leave such things in; was probably useful at some point when developing the test.

File: core/src/test/java/hudson/util/QuotedStringTokenizerTest.java [hunk 4]
```diff
@@ -23,54 +23,63 @@
...
+
 /**
  * @author Kohsuke Kawaguchi
  */
-public class QuotedStringTokenizerTest extends TestCase {
-    private void check(String src, String... expected) {
-        String[] r = QuotedStringTokenizer.tokenize(src);
-        System.out.println(Arrays.asList(r));
-        assertTrue(Arrays.equals(r, expected));
-    }
```
Comment (jglick) [inline core/src/test/java/hudson/util/QuotedStringTokenizerTest.java]: Can minimize diff length by leaving this where it was.

File: core/src/test/java/hudson/util/io/ZipArchiverTest.java [hunk 5]
```diff
@@ -31,20 +33,21 @@ protected void setUp() {
...
         }
     }
-    
-    @Override
-    protected void tearDown() {
+
+    @After
+    public void tearDown() {
         deleteDir(tmpDir);
     }
```
Comment (jglick) [inline core/src/test/java/hudson/util/io/ZipArchiverTest.java]: All better replaced with `@Rule public TemporaryFolder temp;`.

**Comment by jglick** [comment]
I have done some such conversions, mainly to replace `HudsonTestCase` with `JenkinsRule` (which is a little trickier in some cases).

File: core/src/test/java/hudson/MarkupTextTest.java (hunk 2 above)
Comment (darxriggs) [inline core/src/test/java/hudson/MarkupTextTest.java]: I didn't want to change too much at one time.

File: core/src/test/java/hudson/util/io/ZipArchiverTest.java (hunk 5 above)
Comment (darxriggs) [inline core/src/test/java/hudson/util/io/ZipArchiverTest.java]: I didn't want to change too much at one time.

**Comment by darxriggs** [comment]
I have a big PR in the pipeline converting test cases from using ``HudsonTestCase`` to ``JenkinsRule``. Stay tuned. ;-)

Any recommendation on how to split converting 50+ test classes? One commit per class so they can also be cherry-picked?

File: core/src/test/java/hudson/MarkupTextTest.java (hunk 2 above)
Comment (jglick) [inline core/src/test/java/hudson/MarkupTextTest.java]: Sure, just a note for any future PRs you want to do. (Thanks for this work BTW!)

**Comment by jglick** [comment]
A bulk conversion to `JenkinsRule` would be awesome. I would recommend you look around in history for commits I have done along these lines (usually one or two tests at a time) to see the kinds of changes you might need to make.

One commit overall is fine, or multiple commits if you find it more convenient. There is (generally) no need to cherry-pick conversions to JUnit 4. My comment about avoiding whitespace-only changes was about cherry-picking _other_ commits (especially `lts-candidate` bug fixes): if there was an earlier WS-only change on a nearby line, that can cause `git-cherry-pick` on the bug fix commit to fail with a merge conflict, requiring manual (and error-prone) fixup by someone else down the line (such as @olivergondza doing an LTS release). Granted, this is a smaller issue in `src/test/` code, but still an annoyance.

**Comment by jglick** [comment]
Tip:

    git log --grep 'JUnit 4' --stat

71469cea8530310818847b8ae85db4a3e8268200 and 2b0bd977f38c919f19caf4925124ba00523191e7 are reasonably representative.

**Comment by darxriggs** [comment]
Do you realy want to have ``@Rule`` and ``@Test`` in the same line?

**Comment by olivergondza** [comment]
I backport such test case conversions routinely whenever it causes conflicts. Having all this in few commits would help to cherry-pick it early in next LTS line.

**Comment by jglick** [comment]
> Do you really want to have `@Rule` and `@Test` in the same line?

You mean, `@Rule` or `@Test` on the same line as the annotated element. This is my convention: specifically that the one annotation (if any) used as a determiner should be inline, including `@Rule`, `@Test`, `@Override`, `@Extension`, `@WebMethod`, and a few other cases. Most annotations are adjectival (`@Deprecated`, `@SuppressWarnings`, `@Restricted`, etc.), which I would put on their own preceding lines.

Of course like any matter of style, people will vehemently disagree. Use whatever you are comfortable with; there is no particular style guide for Jenkins code.

**Pull request merged**